concordium-consensus/src/Concordium/GRPC2.hs

```diff
@@ -535,11 +563,714 @@ instance ToProto AccountInfo where
+
+instance ToProto DurationSeconds where
+  type Output DurationSeconds = Proto.DurationSeconds
+  toProto = mkWord64
+
+instance ToProto ArInfo.ArInfo where
+  type Output ArInfo.ArInfo = Proto.ArInfo
+  toProto ai = Proto.make $ do
+    ProtoFields.identity .= mkWord32 (ArInfo.arIdentity ai)
+    ProtoFields.description .= Proto.make (do
+      ProtoFields.name .= ArInfo.arName ai
+      ProtoFields.url .= ArInfo.arUrl ai
+      ProtoFields.description .= ArInfo.arDescription ai)
+    ProtoFields.publicKey . ProtoFields.value .= ArInfo.arPublicKey ai
+
+instance ToProto IpInfo.IpInfo where
+  type Output IpInfo.IpInfo = Proto.IpInfo
+  toProto ii = Proto.make $ do
+    ProtoFields.identity .= mkWord32 (IpInfo.ipIdentity ii)
+    ProtoFields.description .= Proto.make (do
+      ProtoFields.name .= IpInfo.ipName ii
+      ProtoFields.url .= IpInfo.ipUrl ii
+      ProtoFields.description .= IpInfo.ipDescription ii)
+    ProtoFields.verifyKey . ProtoFields.value .= IpInfo.ipVerifyKey ii
+    ProtoFields.cdiVerifyKey . ProtoFields.value .= IpInfo.ipCdiVerifyKey ii
+
+instance ToProto Updates.Level1Update where
+  type Output Updates.Level1Update = Proto.UpdatePayload'Level1UpdatePayload
+  toProto Updates.Level1KeysLevel1Update{..} = Proto.make $ ProtoFields.level1KeysUpdate .= toProtoKeys l1kl1uKeys
+  toProto Updates.Level2KeysLevel1Update{..} = Proto.make $ ProtoFields.level2KeysUpdateV0 .= toProtoAuthorizationsV0 l2kl1uAuthorizations
+  toProto Updates.Level2KeysLevel1UpdateV1{..} = Proto.make $ ProtoFields.level2KeysUpdateV1 .= Proto.make (do
+        ProtoFields.v0 .= toProtoAuthorizationsV0 l2kl1uAuthorizationsV1
+        case Updates.asCooldownParameters l2kl1uAuthorizationsV1 of
+            JustForCPV1 as -> ProtoFields.parameterCooldown .= toProto as
+        case Updates.asTimeParameters l2kl1uAuthorizationsV1 of
+            JustForCPV1 as -> ProtoFields.parameterTime .= toProto as)
+
+instance ToProto Updates.RootUpdate where
+  type Output Updates.RootUpdate = Proto.UpdatePayload'RootUpdatePayload
+  toProto ru = case ru of
+    Updates.RootKeysRootUpdate{..} -> Proto.make $ ProtoFields.rootKeysUpdate .= toProtoKeys rkruKeys
+    Updates.Level1KeysRootUpdate{..} -> Proto.make $ ProtoFields.level1KeysUpdate .= toProtoKeys l1kruKeys
+    Updates.Level2KeysRootUpdate{..} -> Proto.make $ ProtoFields.level2KeysUpdateV0 .= toProtoAuthorizationsV0 l2kruAuthorizations
+    Updates.Level2KeysRootUpdateV1{..} -> Proto.make $ ProtoFields.level2KeysUpdateV1 .= Proto.make (do
+        ProtoFields.v0 .= toProtoAuthorizationsV0 l2kruAuthorizationsV1
+        case Updates.asCooldownParameters l2kruAuthorizationsV1 of
+            JustForCPV1 as -> ProtoFields.parameterCooldown .= toProto as
+        case Updates.asTimeParameters l2kruAuthorizationsV1 of
+            JustForCPV1 as -> ProtoFields.parameterTime .= toProto as)
+
+-- |Is not needed if we can make a ToProto instance for HigherLevelKeys.
```

Comment:
I don't quite understand why we are not doing that? Is it because there is a type parameter to the higher-level keys?

Reply:
I recommend just having an instance.